poglink/cogs/rates.py

```diff
@@ -141,13 +146,16 @@ async def compare_posted_rates(self, webpage_url, output_path):
         # compare rates to last rates
         rates_diff = last_rates.get_diff(rates)
 
-        if rates_diff.items:
-            # save rates to file
-            try:
-                with open(output_path, "w+") as f:
-                    json.dump(rates.to_dict(), f, indent=4)
-            except Exception as e:
-                raise RatesWriteError(e) from e
+        try:
+            if rates_diff.items:
+                # save rates to file
+                try:
+                    with open(output_path, "w+") as f:
+                        json.dump(rates.to_dict(), f, indent=4)
+                except Exception as e:
+                    raise RatesWriteError(e) from e
+        except Exception as e:
```

Comment:
Added this to catch when `rates_diff.items` is empty, though I expect this to be caught by an upstream exception in most cases.

Reply:
Hm, could you maybe walk me through the logic of the change here a bit more in the interest of careful review? Just a couple concerns jumping out at me from first glance:

1. If `rates_diff.items` is empty (i.e. an empty list, `[]`), then the `if rates_diff.items:` check should evaluate to False and this statement should never execute. So there's no code that should actually be executing in this block to raise an exception at all.
2. The code that actually is executed in this `if` statement is already wrapped in a `try/catch` block, which means it should already raise an exception as needed.
3. If the interior `open()` statement raises an exception, it's caught by the inner `try/catch` and re-raised as a `RatesWriteError`. By wrapping this entire `if` block around a new `try/catch`, you'll just be catching the newly raised `RatesWriteError` exception again, then raising it again.

So overall, just wondering what problem this was intended to resolve? Maybe it was a red herring and something else was happening? I just can't wrap my head around this block being a problem area.

Reply:
The error that prompted me to add an exception here was 
```AttributeError: 'coroutine' object has no attribute 'items'```

It's not that `rates_diff.items` was empty, but rather `rates_diff` didn't exist at all. This is because its created by the output of `compare_posted_rates`, however [`compare_posted_rates`](https://github.com/FM-17/poglink/blob/cb737672020f580bfdca05aed4c263d21b4648dc/poglink/cogs/rates.py#L170) wasn't awaited in `rates.py`
```python
rates_diff = await self.compare_posted_rates(url, output_path)
```

I was hoping to catch the scenario where an `AttributeError` is caused by `rates_diff` not existing, but I now see the issue outlined in your third point. Please advise if you have a more functional approach in mind, otherwise I'll remove my exceptions for the time being.

Reply:
Removed `RatesReadError` in favor of #44